### PR TITLE
Compare inaccurate scans

### PR DESCRIPTION
**Pull request by prashant905**
# Description

The branch implements the functionality of comparing the inaccurate scans across two models. It finds out the common scans across two models which have greater error margin. 

User story #882  (https://dev.azure.com/cgmorg/ChildGrowthMonitor/_workitems/edit/882/

# How Has This Been Tested?

Functionality test is been tested manually and asserted with successful completion of the pipeline.with output as png image

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests
- [x] New and existing unit tests pass locally with my changes
- [x] My Code is review by 1 People

### CODE REVIEW

File: src/common/evaluation/QA/eval_depthmap_models/combine_results.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 from glob2 import glob
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/combine_results.py:6]: Please revert the unwanted changes to this file

File: src/common/evaluation/QA/environment.yml [hunk 2]
```diff
@@ -20,3 +20,4 @@ dependencies:
     - bunch==1.0.1
     - cgmzscore==2.0.3
     - scikit-learn==0.22.2.post1
+    - matplotlib-venn
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/environment.yml:23]: Can we do without this library somehow? If we really need it then other requirements.txt files might have to be adjusted, but I would prefer plain matplotlib or writing text/table instead

File: src/common/evaluation/QA/eval_depthmap_models/eval_configs.sh [hunk 3]
```diff
@@ -2,9 +2,10 @@
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/eval_configs.sh:2]: This file shouldn't have to change I think. I think you could revert your changes

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import logging
+from typing import List
+
+from glob2 import glob
+from matplotlib_venn import venn2   
+from matplotlib import pyplot as plt
+from pathlib import Path
+
+
+ACC = 2
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:12]: Can you rename this to ACCURACY_THRESHOLD ?

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+FIGURE_NAME = 'common_inaccurate_scans.png'
+
+def merge_qrc(row):
+    """
+    Function to combine qrcodes with scantypes
+    """
+    scans = str(row['qrcode'])+'_'+str(row['scantype'])
+    return scans
+
+def filter_scans(dataframe: pd.DataFrame,accuracy: int):
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:25]: Please type-annotate the return type as well if you annotate the parameter types

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    Function that filter dataframe for the fiven accuracy number
+    """
+    error = dataframe[(dataframe['error'] >= accuracy) | (dataframe['error'] <= -accuracy)]
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:29]: You can work with `abs(dataframe['error'])` to only need one comparison

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    return error
+
+def frame_to_set(dataframe: pd.DataFrame):
+    """
+    Function to convert dataframe column to list
+    """
+    return set(dataframe['name'].to_list())
+
+def calculate_union(set1:set,set2:set):
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:38]: What is the benefit of this function over just calling union directly?

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    
+    csv_files = glob(CSV_PATH)
+    if len(csv_files) !=2:
+        logging.warning("path contains 0 or more than 2 csv files")
+    scan_sets = []
+    for file in csv_files:
+        frame_set = inaccurate_scans(file)
+        scan_sets.append(frame_set)
+        
+    Union_set = calculate_union(scan_sets[0],scan_sets[1])
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:87]: please lowercase local variables

**Review by markus-hinsche — CHANGES_REQUESTED** [review]
Please go through the checklist in the PR description and check what you have done. 
The checks should pass and you should self-review the code. 
Please assign a reviewer again after this

File: src/common/evaluation/QA/environment.yml (hunk 2 above)
Comment (prashant905) [reply src/common/evaluation/QA/environment.yml]: Removed this library and stick to simple report for now.

File: src/common/evaluation/QA/eval_depthmap_models/eval_configs.sh (hunk 3 above)
Comment (prashant905) [reply src/common/evaluation/QA/eval_depthmap_models/eval_configs.sh:2]: changesm reverted according to main branch

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py (hunk 5 above)
Comment (prashant905) [reply src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py]: Done

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py (hunk 6 above)
Comment (prashant905) [reply src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py]: Done

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py (hunk 4 above)
Comment (prashant905) [reply src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py]: Done

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py (hunk 7 above)
Comment (markus-hinsche) [reply src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py]: This comment hasn't been addressed. What are your thoughts? @prashant905

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+def filter_scans(dataframe: pd.DataFrame, accuracy: int) -> pd.DataFrame:
+    """
+    Function that filter dataframe for the fiven accuracy number
+    """
+    error = dataframe[dataframe['error'].abs() >= accuracy]
+    return error
+
+
+def frame_to_set(dataframe: pd.DataFrame) -> pd.DataFrame:
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:29]: The return type looks wrong. I thing this should be a set

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+def extract_model_name(path_name) -> str:
+    """
+    Function to extract the model name from the path.
+    """
+    assert path_name.endswith('.csv')
+    model_name = Path(path_name).resolve().stem
+    return model_name
+
+
+def inaccurate_scans(file) -> pd.DataFrame:
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:61]: Return type looks wrong

Parameter file is missing. 

`file` is a Python keyword, therefore the parameter could be renamed

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    frame_set = frame_to_set(accuracy_df)
+    return frame_set
+
+
+if __name__ == "__main__":
+    csv_files = glob(CSV_PATH)
+    if len(csv_files) != 2:
+        logging.warning("path contains 0 or more than 2 csv files")
+    scan_sets = []
+    for file in csv_files:
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:78]: `file` is a Python keyword, therefore the variable could be renamed

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        frame_set = inaccurate_scans(file)
+        scan_sets.append(frame_set)
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py:80]: [nit] This for loop could benefit from using list comprehension

File: src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py [hunk 13]
```diff
@@ -293,6 +293,11 @@ def get_prediction(model_path: str, dataset_evaluation: tf.data.Dataset) -> np.a
     print(f"Calculate and save the results to {csv_file}")
     utils.calculate_and_save_results(df_grouped, EVAL_CONFIG.NAME, csv_file,
                                      DATA_CONFIG, RESULT_CONFIG, fct=calculate_performance)
+
+#     accuracy_df = utils.filter_scans(df_grouped,ACC)
+    sample_csv_file = f"{OUTPUT_CSV_PATH}/{RUN_ID}_scans.csv"
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py:298]: Please prefix with `scans_` to be consistent

```suggestion
    sample_csv_file = f"{OUTPUT_CSV_PATH}/scans_{RUN_ID}.csv"
```

File: src/common/evaluation/QA/eval_depthmap_models/src/qa_config_filter.py [hunk 14]
```diff
@@ -33,7 +33,7 @@
     CLUSTER_NAME="gpu-cluster",
 
     # Used for Debug the QA pipeline
-    DEBUG_RUN=False,
+    DEBUG_RUN=True,
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/qa_config_filter.py:36]: Please revert this change before merging

File: src/common/evaluation/QA/test-pipeline.yml [hunk 15]
```diff
@@ -47,15 +47,25 @@ jobs:
...
+#   - bash: |
+#       set -euox pipefail
+#       cd src/common/evaluation/QA/eval_depthmap_models
+#       source /usr/share/miniconda/etc/profile.d/conda.sh
+#       conda activate CGM_QA_Pipeline
+#       python combine_results.py --model_measurement height
+#       ls -l outputs/
+#       ls -l
+#     displayName: 'Combine results'
+
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/test-pipeline.yml:59]: Why do we uncomment this?

File: src/data_utils/dataset_EDA/anon-depthmap-95k/EDA_anon-depthmap-95k.ipynb [hunk 16]
```diff
@@ -2,10 +2,10 @@
  "cells": [
```
Comment (markus-hinsche) [inline src/data_utils/dataset_EDA/anon-depthmap-95k/EDA_anon-depthmap-95k.ipynb:2]: Why  are there changes to the EDA notebooks?

**Review by markus-hinsche — CHANGES_REQUESTED** [review]
Too many questionmarks still. A lot of places got changed (e.g. EDA notebooks), but I don't understand why

File: src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py (hunk 7 above)
Comment (prashant905) [reply src/common/evaluation/QA/eval_depthmap_models/inaccurate_scans.py]: not much thought. trying to make code more modular(maybe overdoing it). Makes it extendable and more readable.  Seen this type implementation of making function even for the simplest of functionality.

File: src/common/evaluation/QA/test-pipeline.yml (hunk 15 above)
Comment (prashant905) [reply src/common/evaluation/QA/test-pipeline.yml:59]: Didn't need this part pipeline. hence commented it. revert back according to the master branch.

File: src/data_utils/dataset_EDA/anon-depthmap-95k/EDA_anon-depthmap-95k.ipynb (hunk 16 above)
Comment (prashant905) [reply src/data_utils/dataset_EDA/anon-depthmap-95k/EDA_anon-depthmap-95k.ipynb:2]: make some changes in the EDA notebook to get some data for MS paper. did not push those changes and EDA notebook changed back according to the master branch now.

Comment (markus-hinsche) [reply src/data_utils/dataset_EDA/anon-depthmap-95k/EDA_anon-depthmap-95k.ipynb:2]: I still see changes to EDA notebooks in this PR

**Review by markus-hinsche — COMMENTED** [review]
Almost ready for merge. Now there are a few comments mostly about naming things so this can be understood by a reader who is new to this topic

**Review by markus-hinsche — APPROVED** [review]
LGTM

**Pull request merged**